File: competitions/101-Hack-52/car-show.py
#https://www.hackerrank.com/contests/101hack52/challenges/car-show

import sys
A = 1000001

if __name__ == "__main__":
    n, q = [int(_) for _ in input().strip().split()]
    models = [int(_) for _ in input().strip().split()]
    fur = [0 for i in range(n)]
    seen = [-1 for i in range(A)]
    s = 0
    f = 0
    while s<n:
        if f<n and seen[models[f]] == -1:
            fur[s] = f-s+1
            seen[models[f]] = f
            f += 1
        elif f<n and seen[models[f]] != -1:
            seen[models[s]] = -1
            s += 1
            while s <= seen[models[f]]:
                fur[s] = fur[s-1] -1
                seen[models[s]] = -1
                s += 1
            seen[models[f]] = f
            fur[s] = f-s+1
            f += 1
        elif f == n and s+1 != n:
            fur[s+1] = fur[s]-1
            s += 1
        else:
            s += 1

    for i in range(q):
        l, r = [int(_) for _ in input().split()]
        total = 0
        for j in range(l-1, r):
            total += min(fur[j], r-j)
        print(total)

File: competitions/101-Hack-52/car-show.cpp
//https://www.hackerrank.com/contests/101hack52/challenges/car-show

#include<iostream>
#include<vector>
#include<cassert>
using namespace std;

int A = 1000001;

int main() {
	int n, q, v, l, r;
	long tr, t;
	cin >> n >> q;
	int *models = new int[n];
	int *fur = new int[n];
	int *seen = new int[A];
	for (int i=0; i<A; i++) { seen[i] =0;}
	int *prev = new int[A];
	for (int i=0; i<n; i++) {
		cin >> v;
		models[i] = v;
	}

	int s = 0; int f = 0;
	while (s < n) {
		assert (s<=f);
		cout << s << " " << f;
		if (f < n && seen[models[f]] == 0) {
			fur[s] += 1;
			prev[models[f]] = f;
			seen[models[f]] = 1;
			f += 1;
		}
		else if (f<n && seen[models[f]] == 1) {
			seen[models[f]] = 0;
			s += 1;
			assert(prev[models[f]] >= s)
			while (s != prev[models[f]]) {
				fur[s] = fur[s-1] - 1;
				seen[models[s]] = 0;
				s += 1;
			}
		}
		else if (f == n && s+1 != n) {
			fur[s+1] = fur[s] - 1;
			s += 1;
		}
		else {
			s += 1;
		}
	}

	for (int i=0; i<q; i++) {
		cin >> l >> r;
		t=0;
		for (int j=l; j<=r; j++) {
			tr = fur[j]*(fur[j]+1)/2;
			t += tr;
		}
		cout << t;
	}
	return 0;
}


